server/src/main/java/org/opensearch/crypto/CryptoManagerRegistry.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.crypto;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.cluster.metadata.CryptoMetadata;
+import org.opensearch.common.SetOnce;
+import org.opensearch.plugins.CryptoPlugin;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * During node bootstrap, installed key provider extensions responsible for generating data keys are loaded.
+ * Crypto factories against the respective extensions are cached. A crypto factory is used to register crypto
+ * manager against an {@link org.opensearch.common.blobstore.EncryptedBlobStore}
+ */
+public class CryptoManagerRegistry {
+    private static final Logger logger = LogManager.getLogger(CryptoManagerRegistry.class);
+    // Package private for tests
+    static SetOnce<Map<String, CryptoManager.Factory>> registry = new SetOnce<>();
+    private static final Map<String, CryptoManager> registeredCryptoManagers = new HashMap<>();
+
+    /**
+     * Initializes the registry with crypto factories for the installed crypto key providers.
+     *
+     * @param cryptoPlugins The list of installed crypto plugins containing key providers.
+     */
+    public static void initRegistry(List<CryptoPlugin> cryptoPlugins) {
+        registry.set(loadCryptoFactories(cryptoPlugins));
+    }
+
+    private static Map<String, CryptoManager.Factory> loadCryptoFactories(List<CryptoPlugin> cryptoPlugins) {
+        Map<String, CryptoManager.Factory> cryptoFactories = new HashMap<>();
+        for (CryptoPlugin cryptoPlugin : cryptoPlugins) {
+            for (String keyProviderType : cryptoPlugin.getKeyProviderTypes()) {
+                if (cryptoFactories.containsKey(keyProviderType)) {
+                    throw new IllegalArgumentException("Crypto plugin key provider type [" + keyProviderType + "] is already registered");
+                }
+                cryptoFactories.put(keyProviderType, cryptoPlugin.createClientFactory(keyProviderType));
+            }
+        }
+
+        return cryptoFactories;
```

Comment:
Better to use an immutable map here:

```suggestion
        return Map.copyOf(cryptoFactories);
```

Reply:
done